### ISSUE TITLE
Waypoint UI losing application.log history after changing tabs

### ISSUE THREAD

**Issue by kfh** [hashicorp/waypoint#930]
I noticed that after application startup when following  the _application.log_ in the UI, if i switch to another tab in the ui, ie _Builds_ or _Deployments_  and then go back to the _Logs_ tab immediately the log history is lost. I no longer see the logs from my application startup. But instead i see the following two lines:
```
2020-12-17T13:09:15+01:00: [ERROR] entrypoint.log: log stream disconnected from server, attempting reconnect: err=EOF
2020-12-17T13:08:53+01:00: [ERROR] entrypoint.log: log stream disconnected from server, attempting reconnect: err=EOF
```

For debugging purposes i've also attached an image of how it looks before i switch tabs:

<img width="1595" alt="wp_ui" src="https://user-images.githubusercontent.com/186453/102486929-0fc1c400-406a-11eb-8b2e-60f67ce86792.png">

It should also be mentioned that these _entrypoint.log_ errors where non existant on WP version 0.1.5. For the record, we have been running on AWS Fargate all the time.

**Comment by kfh** [comment]
After having our apps running a couple of weeks now i think there are two _issues_ here: The first one is as mentioned that we are _losing log history_ and the other one is that these _entrypoint.log warnings_ are completely _polluting our logs_. It starts to get a bit frustrating  and if there are any way of disabling all these unneccessary warnings that would be appreciated. @mitchellh Any suggestions?

**Comment by mitchellh** [comment]
I think there is still a UI issue here. To check that: when you "lose" logs on tab switch, if you use `waypoint logs` in the CLI, are they all there? or is that blank too? 

Re: log stream disconnected. Yes, I can see that being annoying. I didn't expect this to be common. It looks like it MIGHT be roughly every ~1minute which seems to hint to me there might be a load balancer killing the connection in the middle. There isn't enough data points to be sure. Regardless, I think that could be an expected behavior so yeah, we should tone that down. Working on it now.

**Comment by mitchellh** [comment]
Re: log stream messages https://github.com/hashicorp/waypoint/pull/953

**Comment by kfh** [comment]
> I think there is still a UI issue here. To check that: when you "lose" logs on tab switch, if you use `waypoint logs` in the CLI, are they all there? or is that blank too?
>

Using `waypoint logs` give me exactly the same behaviour. When i run it just before app deployment i get the full log but when cancel it and re-rerun it almost immediately it just starts from the  _entrypoint.log_ error as mentioned above.

**Comment by mitchellh** [comment]
Both the UI and the CLI only start with the 100 most recent log lines. Do you think you're hitting that limit? Or is this far shorter? (At the API level that is configurable and the server does store more but we don't yet expose that...) I don't _think_ it is this given your original issue but I wanted to double check.

After you see only the two log lines, do you continue to stream in _new_ logs or do all logs stop working?

I have a theory about this being fargate workloads that when the instance deletes we also delete all the logs associated with it and that is what is happening here...

**Comment by kfh** [comment]
> Both the UI and the CLI only start with the 100 most recent log lines. Do you think you're hitting that limit? Or is this far shorter? (At the API level that is configurable and the server does store more but we don't yet expose that...) I don't _think_ it is this given your original issue but I wanted to double check.
> 

This is _far_ shorter so not hitting that limit.

> After you see only the two log lines, do you continue to stream in _new_ logs or do all logs stop working?
> 

From the _command line_ it usually stops after 2 or 3 messages with: `! Error reading logs: rpc error: code = Unavailable desc = transport is closing`

> I have a theory about this being fargate workloads that when the instance deletes we also delete all the logs associated with it and that is what is happening here...

**Comment by mitchellh** [comment]
Yeah okay, this is all pointing towards long-running connections being terminated "early" to the server. The `waypoint logs` command today doesn't reconnect and `code = Unavailable` error is given when the connection is terminated unexpectedly usually from the remote side. 

Are you running the Waypoint server on Fargate as well? Or behind a load balancer perhaps?

I think we need to improve this behavior but my current hunch is this is the root issue.

**Comment by kfh** [comment]
> Yeah okay, this is all pointing towards long-running connections being terminated "early" to the server. The `waypoint logs` command today doesn't reconnect and `code = Unavailable` error is given when the connection is terminated unexpectedly usually from the remote side.
> 
> Are you running the Waypoint server on Fargate as well? Or behind a load balancer perhaps?

The WP server is running on _AWS EKS_ installed using the instructions found in the WP docs.

> 
> I think we need to improve this behavior but my current hunch is this is the root issue.

**Comment by kfh** [comment]
@mitchellh Just a quick update on this issue. I've been upgrading to the latest version of WP(0.2.1) in our infrastructure and all the log spam is gone as shown in the attached image:

<img width="1560" alt="clean" src="https://user-images.githubusercontent.com/186453/106923612-8eec8380-670e-11eb-8376-bdaf1b41e30f.png">

When i switch windows back and fourth after there is no log output at all as shown here:

<img width="1567" alt="empty" src="https://user-images.githubusercontent.com/186453/106923789-bba09b00-670e-11eb-84fd-9ed35f9e0a98.png">

I have been testing this back and fourth with different browsers etc and it's pretty clear to me know. If you don't have the log window open in the ui _before_ the logging code actually runs you will never see it in the WP UI. The same behaviour goes for the _waypoint logs_ command in the cli so those recent 100 log lines expected to be output are effectively _devnulled_. Would be interesting to hear other experiences and see if this is an issues only related to AWS and Fargate.

**Comment by mitchellh** [comment]
Alright cool, one step forward at least. @kfh so just verifying, since you say you can reproduce this with the CLI as well this makes it a core issue potentially so I want to take a look: if you run `waypoint logs` before you `waypoint up`, you never see the logs?

**Comment by kfh** [comment]
> Alright cool, one step forward at least. @kfh so just verifying, since you say you can reproduce this with the CLI as well this makes it a core issue potentially so I want to take a look: if you run `waypoint logs` before you `waypoint up`, you never see the logs?

That is correct.

**Comment by mitchellh** [comment]
I wasn't able to reproduce this with a normal Docker image but I think I remember why now that I'm remembering the context of this issue: the issue is that when the instance is unregistered, the log buffer is deleted. Because you're deploying to a serverless environment, instances get unregistered rather regularly. We probably want to hang on to the logs a bit. 

But locally with a Docker (`waypoint-examples repo: docker/nodejs`) I did the following and it all worked:

1. Run `waypoint init`
2. Run `waypoint logs`, it blocks, keep it running
3. In separate terminal, run `waypoint up` 
4. Logs show up in the first terminal!

**Comment by kfh** [comment]
> I wasn't able to reproduce this with a normal Docker image but I think I remember why now that I'm remembering the context of this issue: the issue is that when the instance is unregistered, the log buffer is deleted. Because you're deploying to a serverless environment, instances get unregistered rather regularly. We probably want to hang on to the logs a bit.
> 
> But locally with a Docker (`waypoint-examples repo: docker/nodejs`) I did the following and it all worked:
> 
> 1. Run `waypoint init`
> 2. Run `waypoint logs`, it blocks, keep it running
> 3. In separate terminal, run `waypoint up`
> 4. Logs show up in the first terminal!

I was a bit quick there and didn't read your question properly. Because what i'm trying to describe is actually the opposite of what you are doing 😄 

Before i describe the case in more detail i wanna give a quick overview of our architecture. I think i've mentioned it in other related issues here but we have a fully automated pipeline, all WP commands are run through Github Actions and branches are continously deployed. When branches are merged to main we also run a release. Most developers do not interact directly with WP only a selected few, myself included, have a WP context installed locally and can run commands against the WP server(which is running on AWS EKS). But all have access to the WP UI. And here the fun starts 😄 

Because the problem i'm trying to describe arises when _after deployments_ a typical user will look into the WP UI and open the application log tab. No trace of the application startup log is to be seen. For this to be seen in the UI we have to open the application log _before the deployment_ is finished. Only then will it be picked up and displayed. So you could say that it's basicly _historical data_ that's not shown. My impression was that you should see the last 100 lines logged independently if it's historical or not.

I had another look at the documentation and it's not really clear if what i'm expecting is supported or not considering the following: _The point of Waypoint logs are to debug emergent application behavior rather than historical behavior. _
But i really hope that what i'm expecting is supported or else the usefulness of the log tools will be of lesser value to us. 

So to sum up, your example has always worked for us but the other way around hasn't. And the behaviour is consistent for both the UI and the _waypoint logs_ command.

**Comment by briancain** [comment]
Hey there! This issue has since been resolved, and I believe since the time this issue was opened, we have overhauled much of the logs feature in Waypoint. If this is still an issue, it would make sense to open a new one for us to investigate. Thanks again!
